src/libraries/System.Private.CoreLib/src/System/Globalization/Normalization.Windows.cs

```diff
@@ -65,84 +73,86 @@ internal static string Normalize(string strInput, NormalizationForm normalizatio
 
             Debug.Assert(strInput != null);
 
-            // we depend on Win32 last error when calling NormalizeString
-            // NormalizeString pinvoke has SetLastError attribute property which will set the last error
-            // to 0 (ERROR_SUCCESS) before executing the calls.
-
-            // Guess our buffer size first
-            int iLength = Interop.Normaliz.NormalizeString((int)normalizationForm, strInput, strInput.Length, null, 0);
+            if (strInput.Length == 0)
+            {
+                return string.Empty;
+            }
 
-            int lastError = Marshal.GetLastWin32Error();
-            // Could have an error (actually it'd be quite hard to have an error here)
-            if ((lastError != Interop.Errors.ERROR_SUCCESS) || iLength < 0)
+            char[]? toReturn = null;
+            try
             {
-                if (lastError == Interop.Errors.ERROR_INVALID_PARAMETER)
+                Span<char> buffer = strInput.Length <= 512
+                    ? stackalloc char[512]
+                    : (toReturn = ArrayPool<char>.Shared.Rent(strInput.Length));
+
+                while (true)
                 {
-                    if (normalizationForm != NormalizationForm.FormC &&
-                        normalizationForm != NormalizationForm.FormD &&
-                        normalizationForm != NormalizationForm.FormKC &&
-                        normalizationForm != NormalizationForm.FormKD)
+                    // we depend on Win32 last error when calling NormalizeString
+                    // NormalizeString pinvoke has SetLastError attribute property which will set the last error
+                    // to 0 (ERROR_SUCCESS) before executing the calls.
+                    int realLength;
+                    unsafe
                     {
-                        throw new ArgumentException(SR.Argument_InvalidNormalizationForm, nameof(normalizationForm));
+                        fixed (char* pInput = strInput)
+                        fixed (char* pDest = &MemoryMarshal.GetReference(buffer))
+                        {
+                            realLength = Interop.Normaliz.NormalizeString(normalizationForm, pInput, strInput.Length, pDest, buffer.Length);
+                        }
                     }
+                    int lastError = Marshal.GetLastWin32Error();
 
-                    throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, nameof(strInput));
+                    switch (lastError)
+                    {
+                        case Interop.Errors.ERROR_SUCCESS:
+                            if (realLength == 0)
+                            {
+                                return string.Empty;
```

Comment:
This is not necessary. The string constructor will handle empty strings for  you like this.

Reply:
I was not aware the ROS<char> ctor does that since it's an internal call, I will remove it